File: java/src/test/java/org/example/GameTest.java
package org.example;

import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GameTest {

    @Test
    void oneCell_shouldDie() {
        Cell cell = new Cell(0, 0);
        Game game = new Game(Set.of(cell));

        game.start(1);
        boolean result = game.isCellAlive(cell);

        assertFalse(result);
    }

    @Test
    void cellWithFourNeighbours_shouldDie() {
        Cell cell = new Cell(1, 1);
        Game game = new Game(Set.of(
                cell,
                new Cell(0, 0),
                new Cell(0, 1),
                new Cell(2, 1),
                new Cell(2, 2)
        ));

        game.start(1);
        boolean result = game.isCellAlive(cell);

        assertFalse(result);
    }

    @Test
    void cellWithExactlyTwoNeighbours_shouldBeAlive() {
        Cell cell = new Cell(1, 1);
        Game game = new Game(Set.of(
                cell,
                new Cell(0, 0),
                new Cell(0, 1)
        ));

        game.start(1);
        boolean result = game.isCellAlive(cell);

        assertTrue(result);
    }

    @Test
    void cellWithExactlyThreeNeighbours_shouldBeAlive() {
        Cell cell = new Cell(1, 1);
        Game game = new Game(Set.of(cell,
                new Cell(0, 0),
                new Cell(0, 1),
                new Cell(2, 2)
        ));

        game.start(1);
        boolean result = game.isCellAlive(cell);

        assertTrue(result);
    }

    @Test
    void emptyCellWithExactlyThreeNeighbours_shouldBeAlive() {
        Cell cell = new Cell(1, 1);
        Game game = new Game(Set.of(
                new Cell(0, 0),
                new Cell(0, 1),
                new Cell(2, 2)
        ));

        game.start(1);
        boolean result = game.isCellAlive(cell);

        assertTrue(result);
    }

    @Test
    void twoCells_shouldAllDie() {
        Cell cell1 = new Cell(0, 0);
        Cell cell2 = new Cell(1, 0);
        Game game = new Game(Set.of(cell1, cell2));

        game.start(1);
        int aliveCells = game.getAliveCells().size();

        assertThat(aliveCells).isZero();
    }

    @Test
    void shouldChangeToDifferentStructureAfter10Iterations() {
        Game game = new Game(Set.of(
                new Cell(0, 0),
                new Cell(0, 1),
                new Cell(-1, 0),
                new Cell(1, 0),
                new Cell(0, -1)
        ));

        game.start(10);
        Set<Cell> aliveCells = game.getAliveCells();

        assertThat(aliveCells)
                .hasSize(12)
                .containsExactlyInAnyOrder(
                        new Cell(3, -1),
                        new Cell(3, 0),
                        new Cell(3, 1),

                        new Cell(-3, -1),
                        new Cell(-3, 0),
                        new Cell(-3, 1),

                        new Cell(-1, 3),
                        new Cell(0, 3),
                        new Cell(1, 3),

                        new Cell(-1, -3),
                        new Cell(0, -3),
                        new Cell(1, -3)
                );
    }
}

File: kotlin/src/test/kotlin/GameTest.kt
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

internal class GameTest {
    @Test
    fun oneCell_shouldDie() {
        val cell = Cell(0, 0)
        val game = Game(setOf(cell))

        game.start(1)
        val result = game.isCellAlive(cell)

        assertFalse(result)
    }

    @Test
    fun cellWithFourNeighbours_shouldDie() {
        val cell = Cell(1, 1)
        val game = Game(
            setOf(
                cell,
                Cell(0, 0),
                Cell(0, 1),
                Cell(2, 1),
                Cell(2, 2)
            )
        )

        game.start(1)
        val result = game.isCellAlive(cell)

        assertFalse(result)
    }

    @Test
    fun cellWithExactlyTwoNeighbours_shouldBeAlive() {
        val cell = Cell(1, 1)
        val game = Game(
            setOf(
                cell,
                Cell(0, 0),
                Cell(0, 1)
            )
        )

        game.start(1)
        val result = game.isCellAlive(cell)

        assertTrue(result)
    }

    @Test
    fun cellWithExactlyThreeNeighbours_shouldBeAlive() {
        val cell = Cell(1, 1)
        val game = Game(
            setOf(
                cell,
                Cell(0, 0),
                Cell(0, 1),
                Cell(2, 2)
            )
        )

        game.start(1)
        val result = game.isCellAlive(cell)

        assertTrue(result)
    }

    @Test
    fun emptyCellWithExactlyThreeNeighbours_shouldBeAlive() {
        val cell = Cell(1, 1)
        val game = Game(
            setOf(
                Cell(0, 0),
                Cell(0, 1),
                Cell(2, 2)
            )
        )

        game.start(1)
        val result = game.isCellAlive(cell)

        assertTrue(result)
    }

    @Test
    fun twoCells_shouldAllDie() {
        val cell1 = Cell(0, 0)
        val cell2 = Cell(1, 0)
        val game = Game(setOf(cell1, cell2))

        game.start(1)
        val aliveCellsCount: Int = game.getAliveCells().size

        assertThat(aliveCellsCount).isZero()
    }

    @Test
    fun shouldChangeToDifferentStructureAfter10Iterations() {
        val game = Game(
            setOf(
                Cell(0, 0),
                Cell(0, 1),
                Cell(-1, 0),
                Cell(1, 0),
                Cell(0, -1)
            )
        )

        game.start(10)
        val aliveCells : MutableSet<Cell> = game.getAliveCells().toMutableSet()

        assertThat(aliveCells)
            .containsExactlyInAnyOrder(
                Cell(3, -1),
                Cell(3, 0),
                Cell(3, 1),

                Cell(-3, -1),
                Cell(-3, 0),
                Cell(-3, 1),

                Cell(-1, 3),
                Cell(0, 3),
                Cell(1, 3),

                Cell(-1, -3),
                Cell(0, -3),
                Cell(1, -3)
            )
    }
}